res/controllers/Reloop-Ready-scripts.js

```diff
@@ -115,239 +251,121 @@ components.Encoder.prototype.inValueScale = function(value) {
-        midi: [0xBE, 0x0A],
-        group: "[Master]",
-        inKey: "gain"
-    });
+        this.rate = new components.Pot({
+            midi: [0xB0 + deckIdx, 0x09], // MSB control: 0x3F
+            inKey: "rate",
+            // tempo fader response: top ("-") @ 0, bottom ("+") @ 1023
+            max: 1023,
+            invert: true,
+            // see explanation of ReloopReady.tempoFaderMiddles for info
+            inValueScale: function(value) {
+                const middle = (this.max + 1) * parent.tempoFaderMiddles[deckIdx];
+                return script.absoluteNonLinInverse(value, 0, middle, this.max, 0, 1);
+            }
+        });
 
-    this.components.cueMix = new components.Pot({
-        midi: [0xBE, 0x0D],
-        group: "[Master]",
-        inKey: "headMix"
-    });
+        this.volume = new components.Pot({
+            midi: [0xB0 + deckIdx, 0x1C],
+            inKey: "volume",
+            // whenever we shift, the component triggers softtakeover.
+            // this is usually not a problem on higher-end hardware,
+            // but on this controller, the fader is so-low resolution
+            // that a quick movement can lead to such a large difference,
+            // that it really does seem like the fader has been moved
+            // while shifting. So softtakeover is engaged, even though
+            // it shouldn't which makes the volume fader unresponsive
+            // in that edge case (this is perceived to happen randomly
+            // to the untrained user and definitely not desirable).
+            // So opt out of any softTakeover-trickery.
+            softTakeover: false,
+        });
 
-    this.components.cueVol = new components.Pot({
-        midi: [0xBE, 0x0C],
-        group: "[Master]",
-        inKey: "headGain"
-    });
+        this.pfl = new components.Button({
+            midi: [0x90 + deckIdx, 0x1B],
+            shiftOffset: -0xD,
+            key: "pfl",
+            type: components.Button.prototype.types.toggle,
+        });
 
-    this.components.browse = new components.ComponentContainer({
-        button: new components.Button({
-            midi: [0x9E, 0x06],
-            unshift: function() {
-                this.group = "[Library]";
-                this.inKey = "GoToItem";
-                this.type = components.Button.prototype.types.push;
-            },
+        this.load = new components.Button({
+            midi: [0x9E, 0x02 + deckIdx],
+            shiftOffset: 0x0D,
             shift: function() {
-                this.group = "[Master]";
-                this.inKey = "maximize_library";
-                this.type = components.Button.prototype.types.toggle;
-            }
-        }),
-        knob: new components.Encoder({
-            midi: [0xBE, 0x00],
-            group: "[Library]",
-            unshift: function() {
-                this.inKey = "MoveVertical";
+                this.inKey = "eject";
+                this.outKey = this.inKey;
```

Comment:
`this.load.shift` is supposed to [`eject`](https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#control-[ChannelN]-eject) a track on the specified deck/channel `deckIdx`, right?

**With the current mapping version, ejecting tracks doesn't work for me, maybe because the related MIDI messages are not set up yet in `Reloop Ready.midi.xml`.**

IMHO the following XML needs to be added to `Reloop Ready.midi.xml`:

```
<!-- ... -->
            <!-- holding shift and pressing left channel's LOAD button to eject track loaded on left channel -->
            <control>
                <group></group>
                <key>ReloopReadyInstance.components.leftChannel.load.input</key>
                <status>0x9E</status>
                <midino>0x1B</midino>
                <options>
                    <script-binding/>
                </options>
            </control>
            <!-- holding shift and UNpressing left channel's LOAD button -->
            <control>
                <group></group>
                <key>ReloopReadyInstance.components.leftChannel.load.input</key>
                <status>0x8E</status>
                <midino>0x1B</midino>
                <options>
                    <script-binding/>
                </options>
            </control>

<!-- ... -->

            <!-- holding shift and pressing right channel's LOAD button to eject track loaded on right channel -->
            <control>
                <group></group>
                <key>ReloopReadyInstance.components.rightChannel.load.input</key>
                <status>0x9E</status>
                <midino>0x10</midino>
                <options>
                    <script-binding/>
                </options>
            </control>
            <!-- holding shift and UNpressing right channel's LOAD button -->
            <control>
                <group></group>
                <key>ReloopReadyInstance.components.rightChannel.load.input</key>
                <status>0x8E</status>
                <midino>0x10</midino>
                <options>
                    <script-binding/>
                </options>
            </control>
<!-- ... -->
```

I haven't understood Mixxx's automagic with respect to [`shiftOffset`](https://github.com/mixxxdj/mixxx/wiki/Components-JS#optional-properties), [`unshift`, and `shift`](https://github.com/mixxxdj/mixxx/wiki/Components-JS#shift-layers). I noticed that the offsets are different for the left-deck-shift-load and the right-deck-shift-load:
* "LEFTLOAD" (`0x02`) versus "SHIFT+LEFTLOAD" (`0x1B`) -> offset is **`0x19`**.
* "RIGHTLOAD" (`0x03`) versus "SHIFT-RIGHTLOAD" (`0x10`) -> offset is **`0x0D`** (as currently used in the code for `shiftOffset`.

I noticed I can comment out `shiftOffset: 0x0D` completely, and the ejecting function still works (as long as the XML part is set up as mentioned before).

Reply:
> I haven't understood Mixxx's automagic with respect to [shiftOffset](https://github.com/mixxxdj/mixxx/wiki/Components-JS#optional-properties), [unshift, and shift](https://github.com/mixxxdj/mixxx/wiki/Components-JS#shift-layers). I noticed that the offsets are different for the left-deck-shift-load and the right-deck-shift-load:

Good catch, I just assumed it was the same. So basically `shift` is called by the enclosing `ComponentContainer` and it reconfigures the component for its shifted behavior, `unshift` is then called on shift-button release (as well as on component init to avoid some code duplication). `shiftOffset` is currently only used for LED-output (as no input binding is done via JS (though #12781 would allow us to change that in the future). 

> this.load.shift is supposed to [eject](https://manual.mixxx.org/2.3/en/chapters/appendix/mixxx_controls.html#control-%5BChannelN%5D-eject) a track on the specified deck/channel deckIdx, right?
> 
> With the current mapping version, ejecting tracks doesn't work for me, maybe because the related MIDI messages are not set up yet in Reloop Ready.midi.xml.

Yup, good catch.